File: svc-position/src/main/java/com/beancounter/position/valuation/AverageCost.java
package com.beancounter.position.valuation;

import com.beancounter.common.model.MoneyValues;
import com.beancounter.common.model.Position;
import com.beancounter.common.model.QuantityValues;
import com.beancounter.common.utils.MathUtils;
import java.math.BigDecimal;
import org.springframework.stereotype.Service;

@Service
public class AverageCost {

  /**
   * Unit cost calculator.
   *
   * @param costBasis accumulation of all trn values that affect cost
   * @param total     current quantity
   * @return unit cost
   */
  public BigDecimal value(BigDecimal costBasis, BigDecimal total) {
    return costBasis.divide(total, MathUtils.getMathContext());
  }

  public void setCostValue(Position position, MoneyValues moneyValues) {
    QuantityValues quantityValues = position.getQuantityValues();
    moneyValues.setCostValue(
        MathUtils.multiply(moneyValues.getAverageCost(), quantityValues.getTotal()));
  }


}

File: svc-position/src/main/kotlin/com/beancounter/position/valuation/AverageCost.kt
package com.beancounter.position.valuation

import com.beancounter.common.model.MoneyValues
import com.beancounter.common.model.Position
import com.beancounter.common.utils.MathUtils.Companion.getMathContext
import com.beancounter.common.utils.MathUtils.Companion.multiply
import org.springframework.stereotype.Service
import java.math.BigDecimal

/**
 * Compute the cost using a simple average.
 */
@Service
class AverageCost {
    /**
     * Unit cost calculator.
     *
     * @param costBasis accumulation of all trn values that affect cost
     * @param total     current quantity
     * @return unit cost
     */
    fun value(
        costBasis: BigDecimal,
        total: BigDecimal?,
    ): BigDecimal {
        return costBasis.divide(total, getMathContext())
    }

    fun getCostValue(
        position: Position,
        moneyValues: MoneyValues,
    ): BigDecimal {
        val quantityValues = position.quantityValues
        return multiply(moneyValues.averageCost, quantityValues.getTotal())!!
    }
}
